### PR TITLE
S1144, S4487: Read symbols only for known names

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
By retrieving symbols only for known names, when analyzing efcore, we reduce the analysis time from ~854678ms to 
~210854ms

Partial fix for #2474

### CODE REVIEW

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Intermediate batch

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 1]
```diff
@@ -103,104 +93,44 @@ public override void VisitAttribute(AttributeSyntax node)
...
-        private void TryStoreFieldAccess(IdentifierNameSyntax node, List<ISymbol> symbols)
-        {
-            var access = ParentAccessType(node);
-            var fieldSymbolUsagesList = GetFieldSymbolUsagesList(symbols);
-            if (HasFlag(access, SymbolAccess.Read))
+            if (HasKnownIdentifier(node))
             {
-                fieldSymbolUsagesList.ForEach(usage => usage.Readings.Add(node));
-            }
+                var symbols = GetSymbols(node).ToList();
```
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:98]: Symbols are iterated more than once and this is the reason why a list or other data structure which is not evaluated when consumed is needed. Returning immutable array will do.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 2]
```diff
@@ -258,17 +193,66 @@ public override void VisitPropertyDeclaration(PropertyDeclarationSyntax node)
...
         /// </summary>
         /// <returns>List of symbols</returns>
-        private IEnumerable<ISymbol> GetSymbols<TSyntaxNode>(TSyntaxNode node, Func<TSyntaxNode, bool> condition)
+        private IEnumerable<ISymbol> GetSymbols<TSyntaxNode>(TSyntaxNode node)
             where TSyntaxNode : SyntaxNode
         {
-            return condition(node)
-                ? GetCandidateSymbols(getSemanticModel(node).GetSymbolInfo(node))
-                : Enumerable.Empty<ISymbol>();
+            return GetCandidateSymbols(getSemanticModel(node).GetSymbolInfo(node));
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:255]: Local function `GetCandidateSymbols` doesn't have to exist anymore after this update.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 3]
```diff
@@ -347,12 +331,38 @@ private AccessorAccess EvaluatePropertyAccesses(ExpressionSyntax node)
                 : AccessorAccess.Get;
         }
 
-        private bool IsKnownIdentifier(SimpleNameSyntax nameSyntax) =>
-            knownSymbolNames.Contains(nameSyntax.Identifier.ValueText);
+        private bool HasKnownIdentifier(SimpleNameSyntax nameSyntax) =>
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:334]: Can you remove this function and embed it's content on call sites? There are just two.

While I was reading the `VisitIdentifierName` the name `HasKnownIdentifier` made an impression that all subnodes are inspected somehow. Even when the function name is good.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 4]
```diff
@@ -103,114 +94,51 @@ public override void VisitAttribute(AttributeSyntax node)
...
+                UsedSymbols.UnionWith(GetSymbols(node));
+            }
+
             base.VisitGenericName(node);
         }
 
         public override void VisitElementAccessExpression(ElementAccessExpressionSyntax node)
         {
-            var symbols = GetSymbols(node, x => true).ToList();
+            var symbols = GetSymbols(node);
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:130]: Can this be pre-filtered with the name-check as well? (maybe in another PR)

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**